src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -75,18 +76,60 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       response.sendError(500, "An error occurred while fetching Youtube Posts");
       return;
     }
+
+    int currentPage = convertToInt(request.getParameter("currentPage"));
+    int postPerPage = convertToInt(request.getParameter("postPerPage"));
+
+    threadTitles.clear();
+    threadSentiments.clear();
+    threadLikes.clear();
+    threadUrls.clear();
     for (YoutubePost post : newPosts) {
       threadTitles.add(post.getTitle());
       threadSentiments.add(analyze.getSentimentScore(post.getContent()));
-      threadUpvotes.add(AnalyzedVideo.getRandomUpvote());
+      threadLikes.add(AnalyzedVideo.getRandomUpvote());
       threadUrls.add(post.getUrl());
+    }
 
-      threadInfoList.put(TITLE, threadTitles);
-      threadInfoList.put(SENTIMENT, threadSentiments);
-      threadInfoList.put(UPVOTES, threadUpvotes);
-      threadInfoList.put(URL, threadUrls);
+    int numOfPages = (int) Math.ceil(threadTitles.size() / postPerPage);
+    createCurrentPage(currentPage, postPerPage);
+    if (numOfPages == 0) {
+      numOfPages++;
     }
+    threadInfoList.put("numOfPages", numOfPages);
+    threadInfoList.put(TITLE, threadTitles);
+    threadInfoList.put(SENTIMENT, threadSentiments);
+    threadInfoList.put(LIKES, threadLikes);
+    threadInfoList.put(URL, threadUrls);
+
     response.setContentType("application/json;");
     response.getWriter().print(threadInfoList);
   }
+  private void createCurrentPage(int currentPage, int postPerPage) {
+    int start = (currentPage - 1) * postPerPage;
+    int end = currentPage * postPerPage;
```

Comment:
Similar to on line 94, you could use[ Math.min](https://docs.oracle.com/javase/7/docs/api/java/lang/Math.html#min(int,%20int)) here to get the smaller of the 2 values.